include/wil/resource.h

```diff
@@ -5241,13 +5241,19 @@ namespace wil
     {
         unique_hglobal_locked() = delete;
 
-        unique_hglobal_locked(_In_ STGMEDIUM& medium) : unique_any<void*, decltype(&::GlobalUnlock), ::GlobalUnlock>(medium.hGlobal)
+        unique_hglobal_locked(HGLOBAL global) : unique_any<void*, decltype(&::GlobalUnlock), ::GlobalUnlock>(global)
         {
             // GlobalLock returns a pointer to the associated global memory block and that's what callers care about.
-            m_globalMemory = GlobalLock(medium.hGlobal);
+            m_globalMemory = GlobalLock(global);
         }
 
-        // In the future, we could easily add additional constructor overloads such as unique_hglobal_locked(HGLOBAL) to make consumption easier.
+        unique_hglobal_locked(const unique_hglobal &global) : unique_hglobal_locked(global.get())

```

Comment:
I would not take a const-reference, because it allows this code
```cpp
wil::unique_hglobal_locked lock{ GetHglobal() };
// HGLOBAL freed by unique_any's destructor, lock.get() will give you invalid memory
```

Instead, it should take a non-const reference, which will force you to keep the HGLOBAL alive:
```cpp
wil::unique_hglobal_locked lock{ GetHglobal() }; // This is a compiler error

auto hglobal{ GetHglobal() };
wil::unique_hglobal_locked lock{ hglobal };
// HGLOBAL not freed until hglobal goes out of scope. lock.get() gives you valid memory
```

Reply:
Yes, that's right, r-values are not welcomed here. ;-)

Reply:
Not sure this is correct. [GlobalLock](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-globallock) says it increments the lock-count, and the object is destroyed when the lock count drops to zero. I think this also means `unique_hglobal_locked` needs to _not_ call `GlobalUnlock` if the `GlobalLock` fails.

So in your example:

```c++
HGLOBAL GetHGlobal();
wil::unique_hglobal_locked lock { GetHGlobal() }; // Success, explicit use of `HGLOBAL` ctor
// what GetHGlobal returned now has lock count +1

auto hglobal { GetHGlobal() }; // Type is HGLOBAL but without a strong reference on it (so "+0")
wil::unique_hglobal_locked lock2 { hglobal }; // Success, explicit use of `HGLOBAL` ctor
// And lock2 is holding is now holding a +1 lock count.  When it goes out of scope hglobal
// still contains what GetHGlobal would have returned; if it was previously locked that
// lock count remains.

Reply:
note, only movable memory has a lock count. new code should use movable as that was a win16 construct not useful on current platforms. Of course there is still code that uses the movable, notably OLE, so we need to support it. but code creating global handles should use `GMEM_FIXED` (`GPTR`)

Reply:
Clipboard related functions are still documented to require movable memory.

Reply:
I have tried it and when GlobalFree is called for locked global memory handle, content of memory is changed (in my case first 8 bytes) and part of memory become inaccessible too (debugger show ?? marks instead of content). Also another call to GlobalLock with original global memory handle has returned NULL. So I think it is not possible to use locked memory after calling GlobalFree because it is deallocated.

```
HGLOBAL mem;
GetHGlobalFromStream(memStream.get(), &mem);
void *m = GlobalLock(mem);  // here I see memory 'm' content OK
GlobalFree(mem);            // after this call memory 'm' is corrupted and partly inaccessible
void *m2 = GlobalLock(mem); // this returns NULL
```

Reply:
`The GlobalFree function will free a locked memory object.`: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-globalfree